pkg/kubelet/kubelet.go

```diff
@@ -2001,18 +2001,22 @@ func (kl *Kubelet) syncLoopIteration(configCh <-chan kubetypes.PodUpdate, handle
 }
 
 // dispatchWork starts the asynchronous sync of the pod in a pod worker.
-// If the pod is terminated, dispatchWork
+// If the pod has completed termination, dispatchWork will perform no action.
 func (kl *Kubelet) dispatchWork(pod *v1.Pod, syncType kubetypes.SyncPodType, mirrorPod *v1.Pod, start time.Time) {
-	if kl.podIsTerminated(pod) {
-		if pod.DeletionTimestamp != nil {
-			// If the pod is in a terminated state, there is no pod worker to
-			// handle the work item. Check if the DeletionTimestamp has been
-			// set, and force a status update to trigger a pod deletion request
-			// to the apiserver.
-			kl.statusManager.TerminatePod(pod)
-		}
+	// check whether we are ready to delete the pod from the API server (all status up to date)
+	podTerminal, containersTerminal, podWorkerTerminal := kl.podAndContainersAreTerminal(pod)
+	if pod.DeletionTimestamp != nil && podTerminal && containersTerminal {
```

Comment:
pod.DeletionTimestamp is returned as true in podTerminal. Is checking for the pod.DeletionTimestamp != nil superfluous here?

Reply:
i think it is the other way around.  `podTerminal` is not needed because if `pod.DeletionTimestamp != nil` then `podTerminal` is `true`.

Reply:
No, success or failed containers are also terminal.

Reply:
Pod is requested for deletion is one type of terminal behavior.  A pod has to be terminal and requested for deletion and all containers are stopped before we signal final deletion

Reply:
I think we can exclude it.